pubsub/psclient/client.go

```diff
@@ -1,97 +1,347 @@
+	}
+
+	return unsubMsg
 }
 
 var defaultTimeout = 10 * time.Second
 
+// Opts defines the psclient Client options.
+type Opts struct {
+	ReadTimeout  time.Duration
+	WriteTimeout time.Duration
+}
+
 // Client wraps a *websocket.Conn.
 type Client struct {
 	*websocket.Conn
-	ReadTimeout  time.Duration
-	WriteTimeout time.Duration
+	ourConn       bool
+	readTimeout   time.Duration
+	writeTimeout  time.Duration
+	reqMtx        sync.Mutex
+	sendMsgChan   chan string
+	recvMsgChan   chan *ClientMessage
+	nextRequestID int64
+	requests      map[int64]chan *pstypes.ResponseMessage
+	responses     map[int64]*pstypes.ResponseMessage
+	sendMtx       sync.Mutex
+	ctx           context.Context
+	shutdown      context.CancelFunc
+}
+
+// New creates a new Client from a URL.
+func New(url string, ctx context.Context, opts *Opts) (*Client, error) {
+	ws, err := websocket.Dial(url, "", "/")
+	if err != nil {
+		return nil, err
+	}
+
+	readTimeout, writeTimeout := defaultTimeout, defaultTimeout
+	if opts != nil {
+		readTimeout = opts.ReadTimeout
+		writeTimeout = opts.WriteTimeout
+	}
+
+	ctx, shutdown := context.WithCancel(ctx)
+	cl := &Client{
+		Conn:         ws,
+		ourConn:      true,
+		readTimeout:  readTimeout,
+		writeTimeout: writeTimeout,
+		sendMsgChan:  make(chan string, 16),
+		recvMsgChan:  make(chan *ClientMessage, 16),
+		requests:     make(map[int64]chan *pstypes.ResponseMessage),
+		responses:    make(map[int64]*pstypes.ResponseMessage),
+		ctx:          ctx,
+		shutdown:     shutdown,
+	}
+
+	go cl.receiver()
+
+	return cl, nil
 }
 
-// New creates a new Client from a *websocket.Conn.
-func New(ws *websocket.Conn) *Client {
+// NewFromConn creates a new Client from a *websocket.Conn.
+func NewFromConn(ws *websocket.Conn, ctx context.Context, opts *Opts) *Client {
 	if ws == nil {
 		return nil
 	}
-	return &Client{
+
+	readTimeout, writeTimeout := defaultTimeout, defaultTimeout
+	if opts != nil {
+		readTimeout = opts.ReadTimeout
+		writeTimeout = opts.WriteTimeout
+	}
+
+	ctx, shutdown := context.WithCancel(ctx)
+	cl := &Client{
 		Conn:         ws,
-		ReadTimeout:  defaultTimeout,
-		WriteTimeout: defaultTimeout,
+		readTimeout:  readTimeout,
+		writeTimeout: writeTimeout,
+		sendMsgChan:  make(chan string, 16),
+		recvMsgChan:  make(chan *ClientMessage, 16),
+		requests:     make(map[int64]chan *pstypes.ResponseMessage),
+		responses:    make(map[int64]*pstypes.ResponseMessage),
+		ctx:          ctx,
+		shutdown:     shutdown,
+	}
+
+	go cl.receiver()
+
+	return cl
+}
+
+// Stop shutsdown the Client. If the websocket connection was created during
+// Client construction, it is is shutdown too.
+func (c *Client) Stop() {
+	log.Trace("Stopping psclient Client...")
+	// Close the user's receive channel.
+	close(c.recvMsgChan)
+
+	// Cancel the Context to stop the receiver loop.
+	c.shutdown()
+
+	// Close the websocket connection.
+	if c.ourConn {
+		if err := c.Conn.Close(); err != nil {
```

Comment:
Depending on which constructor was used, the Client owns the connection (`ourConn`).